service/authorization/v2/cache.go

```diff
@@ -0,0 +1,310 @@
+	)
+	go c.periodicRefresh(ctx)
+
+	return nil
+}
+
+// Stop stops the periodic refresh goroutine if it's running
+func (c *EntitlementPolicyCache) Stop() {
+	// Only attempt to stop the refresh goroutine if an interval was set
+	if c.configuredRefreshInterval > 0 {
+		// Check if stopRefresh is already closed
+		select {
+		case <-c.stopRefresh:
+			// Channel is already closed, nothing to do
+			c.logger.DebugContext(context.Background(), "stop called on already stopped cache")
+			return
+		default:
+			// Channel is still open, proceed with closing
+			// Signal the goroutine to stop
+			close(c.stopRefresh)
+			// Wait with a timeout for the refresh goroutine to complete
+			select {
+			case <-c.refreshCompleted:
+				// Goroutine completed successfully
+			case <-time.After(stopTimeout):
+				// Timeout as a safety mechanism in case the goroutine is stuck
+				c.logger.WarnContext(context.Background(), "timed out waiting for refresh goroutine to complete")
+			}
+		}
+	}
+}
+
+// Refresh manually refreshes the cache by reaching out to policy services. In the event of an error,
+// the cache is marked as not filled, and the error is returned.
+func (c *EntitlementPolicyCache) Refresh(ctx context.Context) error {
+	// Retrieve fresh data from the policy services
+	attributes, err := c.retriever.ListAllAttributes(ctx)
+	if err != nil {
+		return err
+	}
+	subjectMappings, err := c.retriever.ListAllSubjectMappings(ctx)
+	if err != nil {
+		return err
+	}
+	registeredResources, err := c.retriever.ListAllRegisteredResources(ctx)
+	if err != nil {
+		return err
+	}
+
+	// If there is an error when Setting with fresh data, mark not filled so IsReady() will re-attempt refresh
+	err = c.cacheClient.Set(ctx, attributesCacheKey, attributes, authzCacheTags)
+	if err != nil {
+		c.isCacheFilled = false
+		return errors.Join(ErrFailedToSet, err)
+	}
+
+	err = c.cacheClient.Set(ctx, subjectMappingsCacheKey, subjectMappings, authzCacheTags)
+	if err != nil {
+		c.isCacheFilled = false
+		return errors.Join(ErrFailedToSet, err)
+	}
+
+	err = c.cacheClient.Set(ctx, registeredResourcesCacheKey, registeredResources, authzCacheTags)
+	if err != nil {
+		c.isCacheFilled = false
+		return errors.Join(ErrFailedToSet, err)
+	}
+
+	c.logger.DebugContext(ctx,
+		"refreshed EntitlementPolicyCache",
+		slog.Int("attributes_count", len(attributes)),
+		slog.Int("subject_mappings_count", len(subjectMappings)),
+		slog.Int("registered_resources_count", len(registeredResources)),
+	)
+
+	// Mark the cache as filled after a successful refresh
+	c.isCacheFilled = true
+
+	return nil
+}
+
+// ListAllAttributes returns the cached attributes
+func (c *EntitlementPolicyCache) ListAllAttributes(ctx context.Context) ([]*policy.Attribute, error) {
+	var (
+		attributes []*policy.Attribute
+		ok         bool
+	)
+
+	cached, err := c.cacheClient.Get(ctx, attributesCacheKey)
+	if err != nil {
+		if errors.Is(err, cache.ErrCacheMiss) {
+			return attributes, nil
```

Comment:
wouldnt a cache miss indicate a larger error here? -- since the cache is populated on startup the key should always be present. would it be best to return the error instead of nil?

Reply:
I will consider this in a follow up.